File: src/test/java/de/dk/api/codeChallenge/notification/service/NotificationServiceTest.java
package de.dk.api.codeChallenge.notification.service;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import de.dk.api.codeChallenge.notification.model.NotificationTypeEntity;
import de.dk.api.codeChallenge.notification.model.UserEntity;
import de.dk.api.codeChallenge.notification.model.dto.NotificationDto;
import de.dk.api.codeChallenge.notification.model.dto.NotificationResultDto;
import de.dk.api.codeChallenge.notification.model.dto.RegisterRequestDto;
import de.dk.api.codeChallenge.notification.model.dto.UserResponseDto;
import de.dk.api.codeChallenge.notification.model.enums.Category;
import de.dk.api.codeChallenge.notification.repository.NotificationTypeRepository;
import de.dk.api.codeChallenge.notification.repository.UserRepository;

import java.util.*;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class NotificationServiceTest {

    private UserRepository userRepository;
    private NotificationTypeRepository notificationTypeRepository;
    private NotificationTypeProvider notificationTypeProvider;
    private NotificationService service;

    @BeforeEach
    void setup() {
        userRepository = mock(UserRepository.class);
        notificationTypeRepository = mock(NotificationTypeRepository.class);
        notificationTypeProvider = mock(NotificationTypeProvider.class);
        service = new NotificationService(userRepository, notificationTypeRepository, notificationTypeProvider);
    }

    @Test
    void register_givenValidCategory_shouldSaveUserWithAllCategoryTypes() {
        UUID userId = UUID.randomUUID();
        RegisterRequestDto req = new RegisterRequestDto(userId, Set.of("type1"));

        NotificationTypeEntity type1 = new NotificationTypeEntity("type1", Category.A);
        NotificationTypeEntity type2 = new NotificationTypeEntity("type2", Category.A);

        when(notificationTypeRepository.findByNameIgnoreCase("type1")).thenReturn(type1);
        when(notificationTypeProvider.getAllByCategory(Category.A)).thenReturn(List.of(type1, type2));
        when(userRepository.findById(userId)).thenReturn(Optional.empty());
        when(userRepository.save(any())).thenAnswer(invocation -> invocation.getArgument(0));

        UserResponseDto result = service.register(req);

        assertEquals(userId, result.userId());
        assertEquals(Set.of("type1", "type2"), result.notificationTypeNames());
        verify(userRepository, times(1)).save(any());
    }

    @Test
    void register_givenUnknownType_shouldThrowBadRequest() {
        UUID userId = UUID.randomUUID();
        RegisterRequestDto req = new RegisterRequestDto(userId, Set.of("unknown-type"));

        when(notificationTypeRepository.findByNameIgnoreCase("unknown-type")).thenReturn(null);

        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class, () -> service.register(req));
        assertEquals("Unknown notification type(s): unknown-type", ex.getMessage());
    }

    @Test
    void register_givenEmptyNotificationSet_shouldThrowBadRequest() {
        UUID userId = UUID.randomUUID();
        RegisterRequestDto req = new RegisterRequestDto(userId, Set.of());

        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class, () -> service.register(req));
        assertEquals("notificationTypeNames must not be empty", ex.getMessage());
    }

    @Test
    void sendNotification_givenSubscribedUser_shouldReturnSent() {
        UUID userId = UUID.randomUUID();

        NotificationTypeEntity notificationType = new NotificationTypeEntity("type1", Category.A);
        UserEntity userEntity = new UserEntity(userId, new HashSet<>(Set.of(notificationType)));

        when(notificationTypeRepository.findByNameIgnoreCase("type1")).thenReturn(notificationType);
        when(userRepository.findByIdWithNotifications(userId)).thenReturn(Optional.of(userEntity));

        NotificationDto dto = new NotificationDto(userId, "type1", "Hello");

        NotificationResultDto result = service.sendNotification(dto);

        assertEquals("SENT", result.status());
        verify(userRepository, times(1)).findByIdWithNotifications(userId);
        verify(notificationTypeRepository, times(1)).findByNameIgnoreCase("type1");
    }

    @Test
    void sendNotification_givenUnsubscribedUser_shouldReturnSkipped() {
        UUID userId = UUID.randomUUID();

        NotificationTypeEntity notificationType = new NotificationTypeEntity("type1", Category.A);
        UserEntity userEntity = new UserEntity(userId, new HashSet<>());

        when(notificationTypeRepository.findByNameIgnoreCase("type1")).thenReturn(notificationType);
        when(userRepository.findByIdWithNotifications(userId)).thenReturn(Optional.of(userEntity));

        NotificationDto dto = new NotificationDto(userId, "type1", "Hello");

        NotificationResultDto result = service.sendNotification(dto);

        assertEquals("SKIPPED", result.status());
        assertEquals("USER_NOT_SUBSCRIBED_TO_CATEGORY", result.reason());
    }

    @Test
    void sendNotification_givenUnknownType_shouldThrowBadRequest() {
        UUID userId = UUID.randomUUID();

        when(notificationTypeRepository.findByNameIgnoreCase("type1")).thenReturn(null);

        NotificationDto dto = new NotificationDto(userId, "type1", "Hello");

        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class, () -> service.sendNotification(dto));
        assertEquals("Unknown notification type: type1", ex.getMessage());
    }

    @Test
    void sendNotification_givenMissingUser_shouldThrowNotFound() {
        UUID userId = UUID.randomUUID();
        NotificationTypeEntity notificationType = new NotificationTypeEntity("type1", Category.A);

        when(notificationTypeRepository.findByNameIgnoreCase("type1")).thenReturn(notificationType);
        when(userRepository.findByIdWithNotifications(userId)).thenReturn(Optional.empty());

        NotificationDto dto = new NotificationDto(userId, "type1", "Hello");

        NoSuchElementException ex = assertThrows(NoSuchElementException.class, () -> service.sendNotification(dto));
        assertEquals("User not found: " + userId, ex.getMessage());
    }
}

File: src/test/kotlin/de/dkb/api/codeChallenge/notification/service/NotificationServiceTest.kt
package de.dkb.api.codeChallenge.notification.service

import de.dkb.api.codeChallenge.notification.model.NotificationTypeEntity
import de.dkb.api.codeChallenge.notification.model.UserEntity
import de.dkb.api.codeChallenge.notification.model.dto.NotificationDto
import de.dkb.api.codeChallenge.notification.model.dto.RegisterRequestDto
import de.dkb.api.codeChallenge.notification.model.enums.Category
import de.dkb.api.codeChallenge.notification.repository.NotificationTypeRepository
import de.dkb.api.codeChallenge.notification.repository.UserRepository
import io.mockk.*
import org.junit.Assert.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.*
import kotlin.test.assertFailsWith

class NotificationServiceTest {

    private lateinit var userRepository: UserRepository
    private lateinit var notificationTypeRepository: NotificationTypeRepository
    private lateinit var cacheService: NotificationTypeCacheService
    private lateinit var service: NotificationService

    @BeforeEach
    fun setup() {
        userRepository = mockk()
        notificationTypeRepository = mockk()
        cacheService = mockk()
        service = NotificationService(userRepository, notificationTypeRepository, cacheService)
    }

    @Test
    fun register_givenValidCategory_shouldSaveUserWithAllCategoryTypes() {

        val userId = UUID.randomUUID()
        val req = RegisterRequestDto(userId, setOf("type1"))

        val type1 = NotificationTypeEntity("type1", Category.A)
        val type2 = NotificationTypeEntity("type2", Category.A)

        every { notificationTypeRepository.findAllByCategory(Category.A) } returns listOf(type1)
        every { cacheService.getAllByCategory(Category.A) } returns listOf(type1, type2)
        every { userRepository.findById(userId) } returns Optional.empty()
        every { userRepository.save(any()) } answers { firstArg() }

        val result = service.register(req)

        assertEquals(userId, result.userId)
        assertEquals(setOf("type1", "type2"), result.notificationTypeNames)
        verify(exactly = 1) { userRepository.save(any()) }
    }

    @Test
    fun sendNotification_givenSubscribedUser_shouldLogSend() {

        val userId = UUID.randomUUID()
        val notificationType = NotificationTypeEntity("type1", Category.A)
        val userEntity = UserEntity(id = userId, notifications = mutableSetOf(notificationType))

        every { notificationTypeRepository.findByNameIgnoreCase("type1") } returns notificationType
        every { userRepository.findByIdWithNotifications(userId) } returns Optional.of(userEntity)

        val dto = NotificationDto(userId, "type1", "Hello")

        service.sendNotification(dto)

        verify(exactly = 1) { userRepository.findByIdWithNotifications(userId) }
        verify(exactly = 1) { notificationTypeRepository.findByNameIgnoreCase("type1") }
    }

    @Test
    fun sendNotification_givenUnsubscribedUser_shouldThrowOrSkip() {
        val userId = UUID.randomUUID()
        val notificationType = NotificationTypeEntity("type1", Category.A)
        val userEntity = UserEntity(id = userId, notifications = mutableSetOf())

        every { notificationTypeRepository.findByNameIgnoreCase("type1") } returns notificationType
        every { userRepository.findByIdWithNotifications(userId) } returns Optional.of(userEntity)

        val dto = NotificationDto(userId, "type1", "Hello")

        assertFailsWith<IllegalArgumentException> {
            service.sendNotification(dto)
        }

        verify(exactly = 1) { userRepository.findByIdWithNotifications(userId) }
    }
}
